libcst/matchers/_matcher_base.py

```diff
@@ -1611,3 +1613,198 @@ def extractall(
+
+    def _node_translate(
+        self, node_or_sequence: Union[libcst.CSTNode, Sequence[libcst.CSTNode]],
+    ) -> Union[libcst.CSTNode, Sequence[libcst.CSTNode]]:
+        if isinstance(node_or_sequence, Sequence):
+            return tuple(self.node_lut[node] for node in node_or_sequence)
+        else:
+            return self.node_lut[node_or_sequence]
+
+    def _extraction_translate(
+        self, extracted: Dict[str, Union[libcst.CSTNode, Sequence[libcst.CSTNode]]]
+    ) -> Dict[str, Union[libcst.CSTNode, Sequence[libcst.CSTNode]]]:
+        return {key: self._node_translate(val) for key, val in extracted.items()}
+
+    def on_leave(
+        self, original_node: libcst.CSTNode, updated_node: libcst.CSTNode
+    ) -> Union[libcst.CSTNode, MaybeSentinel, RemovalSentinel]:
+        # Track original to updated node mapping for this node.
+        self.node_lut[original_node] = updated_node
+
+        # This gets complicated. We need to do the match on the original node,
+        # but we want to do the extraction on the updated node. This is so
+        # metadata works properly in matchers. So, if we get a match, we fix
+        # up the nodes in the match and return that to the replacement lambda.
+        extracted = _matches(original_node, self.matcher, self.metadata_lookup)
+        if extracted is not None:
+            try:
+                # Attempt to do a translation from original to updated node.
+                extracted = self._extraction_translate(extracted)
+            except KeyError:
+                # One of the nodes we looked up doesn't exist anymore, this
+                # is no longer a match. This can happen if a child node was
+                # modified, making this original match not applicable anymore.
+                extracted = None
+        if extracted is not None:
+            # We're replacing this node entirely, so don't save the original
+            # updated node. We don't want this to be part of a parent match
+            # since we can't guarantee that the update matches anymore.
+            del self.node_lut[original_node]
+            return self.replacement(updated_node, extracted)
+        return updated_node
+
+
+def replace(
+    tree: Union[MaybeSentinel, RemovalSentinel, libcst.CSTNode, meta.MetadataWrapper],
+    matcher: Union[
+        BaseMatcherNode, MatchIfTrue[Callable[[object], bool]], _BaseMetadataMatcher,
+    ],
+    replacement: Union[
+        MaybeSentinel,
+        RemovalSentinel,
+        libcst.CSTNode,
+        Callable[
+            [
+                libcst.CSTNode,
+                Dict[str, Union[libcst.CSTNode, Sequence[libcst.CSTNode]]],
+            ],
+            Union[MaybeSentinel, RemovalSentinel, libcst.CSTNode],
+        ],
+    ],
+    *,
+    metadata_resolver: Optional[
+        Union[libcst.MetadataDependent, libcst.MetadataWrapper]
+    ] = None,
+) -> Union[MaybeSentinel, RemovalSentinel, libcst.CSTNode]:
```

Comment:
The over-complicated type annotation make it hard for user to figure out the parameters for `extract()`.
![image](https://user-images.githubusercontent.com/3840867/69991046-cb659380-14fb-11ea-8a14-09485b25c5b2.png)